File: Difficulty: Medium/Majority Element/majority-element.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// User function template for C++
class Solution {
  public:
    int majorityElement(vector<int>& arr) {
        // code here
        //Bayers Moore Majority Algo
        int count = 0;
        int candidate = -1;
        for(auto val : arr)
            {
                if(count ==0)
                    {
                        count = 1;
                        candidate = val;
                    }
                else if(candidate == val)
                    count++;
                else
                    count--;
            }
        count = 0;
        for(auto val : arr)
            {
                if(candidate == val)
                    count++;
            }
        return (count > arr.size()/2)?candidate:-1;
    }
};


//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        int n;
        vector<int> a, b;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        while (ss >> num)
            a.push_back(num);

        Solution obj;
        cout << obj.majorityElement(a) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Majority Element/majority-element.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class Geeks
{
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            int n =sc.nextInt();
            int arr[] = new int[n];
            
            for(int i = 0; i < n; i++)
             arr[i] = sc.nextInt();
             
           System.out.println(new Solution().majorityElement(arr, n)); 
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
     static int majorityElement(int a[], int size)
    {
        int el=0;
        int cnt=0;
        for(int i=0;i<size;i++)
          {
              if(cnt==0)
                {
                    cnt=1;
                    el=a[i];
                }
                else if(el==a[i])
                  cnt++;
                else
                 cnt--;
          }
          cnt=0;
          for(int i=0;i<size;i++)
            {
                if(a[i]==el)
                 cnt++;
            }
            if(cnt> size/2)
              return el;
             else
              return -1;
    }
}